experimental/wbuild-examples/FWF/src/Canvas/Canvas.man
./"" Canvas.3  1.1  Jan 14, 1994
./""  Copyright 1990,1993,1994 Robert Forsman
./""  Gnu Library General Public License version 2.0
./""
./""
.TH Canvas 3 "14 Jan 1994" "Version 3.6" "Free Widget Foundation"
.SH NAME
Canvas - a Widget that allows the programmer to specify a redraw procedure
and a resize procedure on a per-widget basis.

.SH VITAL STATS

.nf
.TA 3.0i
.ta 3.0i
Application Header file		<Canvas.h>
Class Header file		<CanvasP.h>
Class				xfwfcanvasWidgetClass
Class Name			Canvas
Superclass			Core
.fi

  The Canvas widget is useful when the program needs an area for
drawing or display and the creation of a custom widget would not be
``cost-effective''.  Examples of this are display of an XImage,
display of an animation, or display of a custom diagram.  The Canvas
widget is a gain over registering an action proc for an Expose event
because the class definition takes advantage of the X toolkit's
internal expose event compression capabilities to allow the
application to send the drawing commands only once for a collection of
Expose events.

.SH PROGRAMMER INFORMATION

  The Canvas widget is identical to the Core widget except that its
compress_exposure field is XtExposeCompressMultiple and it allows the
programmer to specify both an Expose procedure and a Resize procedure
on per widget basis.

The following resources are specific to the Canvas widget

.TS H
lw(1i) lw(1i) lw(1i) lw(1i).
_
.sp .6
.TB
Name	Class	Type	Default Value
.sp .6
_
visual	Visual	Visual*	CopyFromParent
exposeProc	ExposeProc	CanvExposeProc	NULL
exposeProcData	ExposeProcData	XtPointer	NULL
resizeProc	ResizeProc	CanvResizeProc	NULL
resizeProcData	ResizeProcdata	XtPointer	NULL
.sp .6
_
.TE

.IP \fBvisual\fP
This is used if the canvas must be a different visual from the parent.
This comes in very handy for an 8-bit application that wants to
display a 24-bit image in a subwindow.  The only Intrinsics-supplied
widgets that have an XtNvisual resource are the Shell widgets.
.IP
\fB!Warning!\fP If you specify a visual for the Canvas widget, make
sure that you specify a colormap and depth that match (these resources
are part of the Core widget).  If you do not, you will almost
certainly generate a BadMatch error from the Xlib layer when you
realize the canvas.
.IP \fBexposeProc\fP
This is a pointer to a function of type XfwfCanvasExposeProc which is
identical to XtExposeProc except that it has an extra XtPointer
argument at the end of the argument list through which the
exposeProcData is passed.
.IP \fBexposeProcData\fP
This is a pointer to widget-specific data that will be passed in
through the client_data of the exposeProc.
.IP \fBresizeProc\fP
This is a pointer to a function of type XfwfCanvasResizeProc which is
identical to XtWidgetProc except that it has an extra XtPointer
argument at the end of the argument list through which the
resizeProcData is passed.
.IP \fBresizeProcData\fP
This is a pointer to widget-specific data that will be passed in
through the client_data of the resizeProc.

.SH TYPEDEFS

  The following types are defined in the public header file for the
programmer's enlightenment.

.nf
.TA 1i
typedef void(*XfwfCanvasExposeProc)(
	Widget w,
	XExposeEvent *event,
	Region region,
	XtPointer client_data );

typedef void(*XfwfCanvasResizeProc)(
	Widget w,
	XtPointer client_data );
.fi

.SH COPYRIGHT
Copyright (C) 1990,93,94 Robert Forsman

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the Free
Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

.SH "AUTHOR"
.nf
Robert Forsman
thoth@cis.ufl.edu
http://www.cis.ufl.edu/~thoth
University of Florida
Department of Computer and Information Science
.fi


clean-build/generated/Canvas.c
/* Generated by wbuild
 * (generator version 3.3)
 */
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 156 "widgets/Canvas.widget"
#include <assert.h>
#line 157 "widgets/Canvas.widget"
#include <stdint.h>
#line 158 "widgets/Canvas.widget"
#include <X11/Intrinsic.h>
#line 159 "widgets/Canvas.widget"
#include <X11/Xmu/Converters.h>
#line 160 "widgets/Canvas.widget"
#include <X11/Xft/Xft.h>
#line 161 "widgets/Canvas.widget"
#include <X11/Xregion.h>
#line 162 "widgets/Canvas.widget"
#include "converters.h"
#line 163 "widgets/Canvas.widget"
#include "xutil.h"
#line 164 "widgets/Canvas.widget"
#include "mls.h"
#line 165 "widgets/Canvas.widget"
#include "canvas-draw.h"
#include <xtcw/CanvasP.h>
#line 116 "widgets/Canvas.widget"
static void zoom(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);

static XtActionsRec actionsList[] = {
{"zoom", zoom},
};
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 26 "widgets/Canvas.widget"
static void expose(
#if NeedFunctionPrototypes
Widget,XEvent *,Region 
#endif
);
#line 33 "widgets/Canvas.widget"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 45 "widgets/Canvas.widget"
static void destroy(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 53 "widgets/Canvas.widget"
static void resize(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 61 "widgets/Canvas.widget"
static void realize(
#if NeedFunctionPrototypes
Widget,XtValueMask *,XSetWindowAttributes *
#endif
);
#line 94 "widgets/Canvas.widget"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 101 "widgets/Canvas.widget"
static void paint1(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 107 "widgets/Canvas.widget"
static void init_colors(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 101 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 101 "widgets/Canvas.widget"
static void paint1(Widget self)
#else
#line 101 "widgets/Canvas.widget"
static void paint1(self)Widget self;
#endif
#line 102 "widgets/Canvas.widget"
{
    XTFUNC();
    XtCallCallbackList( self, ((CanvasWidget)self)->canvas.callback, (XtPointer) & ((CanvasWidget)self)->canvas.canv );
}
#line 107 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 107 "widgets/Canvas.widget"
static void init_colors(Widget self)
#else
#line 107 "widgets/Canvas.widget"
static void init_colors(self)Widget self;
#endif
#line 108 "widgets/Canvas.widget"
{
	pixel_to_xftcolor( self, ((CanvasWidget)self)->core.background_pixel, BG );
 	pixel_to_xftcolor( self, ((CanvasWidget)self)->canvas.foreground,       FG );
}

static XtResource resources[] = {
#line 10 "widgets/Canvas.widget"
{XtNxftFont,XtCXFtFont,XtRXftFont,sizeof(((CanvasRec*)NULL)->canvas.xftFont),XtOffsetOf(CanvasRec,canvas.xftFont),XtRString,(XtPointer)"Serif-22"},
#line 11 "widgets/Canvas.widget"
{XtNforeground,XtCForeground,XtRPixel,sizeof(((CanvasRec*)NULL)->canvas.foreground),XtOffsetOf(CanvasRec,canvas.foreground),XtRImmediate,(XtPointer)XtDefaultForeground },
#line 12 "widgets/Canvas.widget"
{XtNcallback,XtCCallback,XtRCallback,sizeof(((CanvasRec*)NULL)->canvas.callback),XtOffsetOf(CanvasRec,canvas.callback),XtRImmediate,(XtPointer)NULL },
#line 13 "widgets/Canvas.widget"
{XtNtopx,XtCTopx,XtRInt,sizeof(((CanvasRec*)NULL)->canvas.topx),XtOffsetOf(CanvasRec,canvas.topx),XtRImmediate,(XtPointer)0 },
#line 14 "widgets/Canvas.widget"
{XtNtopy,XtCTopy,XtRInt,sizeof(((CanvasRec*)NULL)->canvas.topy),XtOffsetOf(CanvasRec,canvas.topy),XtRImmediate,(XtPointer)0 },
#line 15 "widgets/Canvas.widget"
{XtNzoom_thousandth,XtCZoom_thousandth,XtRInt,sizeof(((CanvasRec*)NULL)->canvas.zoom_thousandth),XtOffsetOf(CanvasRec,canvas.zoom_thousandth),XtRImmediate,(XtPointer)1000 },
};

CanvasClassRec canvasClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &coreClassRec,
/* class_name   	*/  "Canvas",
/* widget_size  	*/  sizeof(CanvasRec),
/* class_initialize 	*/  NULL,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  realize,
/* actions      	*/  actionsList,
/* num_actions  	*/  1,
/* resources    	*/  resources,
/* num_resources 	*/  6,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  True ,
/* compress_exposure 	*/  XtExposeCompressMultiple ,
/* compress_enterleave 	*/  True ,
/* visible_interest 	*/  False ,
/* destroy      	*/  destroy,
/* resize       	*/  resize,
/* expose       	*/  expose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  XtInheritTranslations,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* Canvas_class part */
 /* dummy */  0
},
};
WidgetClass canvasWidgetClass = (WidgetClass) &canvasClassRec;
/*ARGSUSED*/
#line 116 "widgets/Canvas.widget"
static void zoom(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
	XTFUNC();
	float scale;
	int parse = 0;
	for( int i=0;i<*num_params;i++)
	{
		char *s = params[i];
		TRACE(1,"PARSE: %s", s );
		switch( parse ) {
			//---- key
			case 0: if( strcasecmp(s, "scale") == 0 ) parse=2;
			        break;
			//---- value
			case 2: TRACE(1,"PARSE");
				scale=strtof(s,NULL);
			        canvas_zoom( self, & ((CanvasWidget)self)->canvas.canv, scale );
				parse=0;
				break;
			default: WARN("uknown parameter: %s", s );
		}
	}
}

static void _resolve_inheritance(class)
WidgetClass class;
{
  CanvasWidgetClass c __attribute__((unused)) = (CanvasWidgetClass) class;
  CanvasWidgetClass super __attribute__((unused));
  if (class == canvasWidgetClass) return;
  super = (CanvasWidgetClass)class->core_class.superclass;
}
#line 26 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 26 "widgets/Canvas.widget"
static void expose(Widget self,XEvent * event,Region  region)
#else
#line 26 "widgets/Canvas.widget"
static void expose(self,event,region)Widget self;XEvent * event;Region  region;
#endif
#line 27 "widgets/Canvas.widget"
{
    XTFUNC();
    /* XftDrawRect(DRAW,BG,0,0,$width,$height); */
    paint1(self);
}
#line 33 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 33 "widgets/Canvas.widget"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 33 "widgets/Canvas.widget"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 34 "widgets/Canvas.widget"
{

    XTFUNC();
    memset(& ((CanvasWidget)self)->canvas.canv, 0, sizeof (((CanvasWidget)self)->canvas.canv));

	if( ((CanvasWidget)self)->core.width <=0 ) ((CanvasWidget)self)->core.width = 20;
	if( ((CanvasWidget)self)->core.height <=0 ) ((CanvasWidget)self)->core.height = 20;
	init_colors(self);
	DRAW=0;
}
#line 45 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 45 "widgets/Canvas.widget"
static void destroy(Widget self)
#else
#line 45 "widgets/Canvas.widget"
static void destroy(self)Widget self;
#endif
#line 46 "widgets/Canvas.widget"
{
        XTFUNC();
        if( DRAW ) XftDrawDestroy( DRAW );
        XtReleaseGC( self, ((CanvasWidget)self)->canvas.canv.gc[0] );
        XtReleaseGC( self, ((CanvasWidget)self)->canvas.canv.gc[1] );
}
#line 53 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 53 "widgets/Canvas.widget"
static void resize(Widget self)
#else
#line 53 "widgets/Canvas.widget"
static void resize(self)Widget self;
#endif
#line 54 "widgets/Canvas.widget"
{
	XTFUNC();
    ((CanvasWidget)self)->canvas.canv.win_width  = ((CanvasWidget)self)->core.width;
    ((CanvasWidget)self)->canvas.canv.win_height = ((CanvasWidget)self)->core.height;

}
#line 61 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 61 "widgets/Canvas.widget"
static void realize(Widget self,XtValueMask * mask,XSetWindowAttributes * attributes)
#else
#line 61 "widgets/Canvas.widget"
static void realize(self,mask,attributes)Widget self;XtValueMask * mask;XSetWindowAttributes * attributes;
#endif
#line 62 "widgets/Canvas.widget"
{
    XTFUNC();
	XtCreateWindow(self, (unsigned int) InputOutput,
        		  (Visual *) CopyFromParent,
			  *mask, attributes);
    Display *dpy = XtDisplay(self);
  	DRAW = XftDrawCreate(dpy, XtWindow(self),
                 DefaultVisual(dpy, DefaultScreen(dpy)), None);

    ((CanvasWidget)self)->canvas.canv.xfont    = ((CanvasWidget)self)->canvas.xftFont;
    ((CanvasWidget)self)->canvas.canv.dpy      =  dpy;
    ((CanvasWidget)self)->canvas.canv.screen   = DefaultScreen(dpy);
    
    XGCValues     values;
    values.graphics_exposures = False;
    values.foreground = ((CanvasWidget)self)->canvas.foreground;
    values.background = ((CanvasWidget)self)->core.background_pixel;

    ((CanvasWidget)self)->canvas.canv.gc[0] = XtGetGC( self,
      (unsigned) GCForeground | GCBackground | GCGraphicsExposures,
      &values);

    values.foreground = ((CanvasWidget)self)->core.background_pixel;
    ((CanvasWidget)self)->canvas.canv.gc[1] = XtGetGC( self,
      (unsigned) GCForeground | GCBackground | GCGraphicsExposures,
      &values);

     ((CanvasWidget)self)->canvas.canv.app_context = XtWidgetToApplicationContext( self );
}
#line 94 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 94 "widgets/Canvas.widget"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 94 "widgets/Canvas.widget"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 95 "widgets/Canvas.widget"
{
	return True; /* yes we want a redraw */	
}
#line 142 "widgets/Canvas.widget"
#line 143 "widgets/Canvas.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 143 "widgets/Canvas.widget"
canvas_draw_t * canvas_get_priv(Widget self)
#else
#line 143 "widgets/Canvas.widget"
canvas_draw_t * canvas_get_priv(self)Widget self;
#endif
#line 144 "widgets/Canvas.widget"
{
    if( XtIsSubclass( self,canvasWidgetClass ) )
      return & ((CanvasWidget)self)->canvas.canv;
    XtWarning( "Widget is not subclass of canvas" );
    return NULL;
}
